### PR TITLE
Implement structural extraction for clauses, tables, and paragraphs (WI-4)

### PR DESCRIPTION
**Pull request by brianluby**
## Summary

Implements comprehensive Markdown structural extraction for list items (clauses), tables, and standalone paragraphs (WI-4), building on the heading extraction from WI-3.

## Features

- ✅ **List Item Extraction**: Extract ordered and unordered list items with nesting depth tracking (up to 255 levels with saturation)
- ✅ **Table Extraction**: Parse GitHub-Flavored Markdown tables with headers and data rows
- ✅ **Paragraph Extraction**: Capture standalone paragraph blocks (not inside lists or tables)
- ✅ **Text Normalization**: Strip inline formatting to plain text (SEC-3)
- ✅ **Source Tracking**: Track source line numbers for all extracted elements (M-4)
- ✅ **Event-Based Parsing**: Use pulldown-cmark's event stream (SEC-6, SEC-7)

### Core Changes
- **src/parse/clauses.rs** (1,000 lines): New module with `extract_clauses()` function and comprehensive test suite
- **src/parse/mod.rs**: Promoted utility functions to `pub(crate)`, added module declaration and re-exports
- **.gitignore**: Added `*.log` pattern

### Architecture
- Single-pass O(n) state machine processing three extraction types concurrently
- Stack-based nesting depth tracking for list items
- State machine for table header/row/cell accumulation
- Standalone paragraph detection (not inside lists or tables)

## Security Requirements Met

- **SEC-1**: Handle empty documents without errors
- **SEC-2**: Empty table cells produce empty strings, no panics
- **SEC-3**: Inline formatting stripped to plain text
- **SEC-4**: Bounded depth with u8 saturation at 255 levels
- **SEC-5**: Source line tracking for all elements
- **SEC-6**: Event-based parsing, no regex
- **SEC-7**: GFM table support via `Options::ENABLE_TABLES`

## Test Coverage

- **92 tests passing** (87 unit tests + 5 doc tests)
- **Zero clippy warnings**
- **Mutation testing ready** with cargo-mutants

### Test Categories
- List extraction (ordered, unordered, nested, mixed types)
- Table extraction (headers, data rows, empty cells)
- Paragraph extraction (standalone, with inline formatting)
- Edge cases (empty documents, deeply nested structures, restarted numbering)
- Integration tests (documents with multiple element types)

## Documentation

Complete specification artifacts in `specs/004-structural-extraction-clauses/`:
- **spec.md**: User stories, acceptance criteria, and requirements
- **plan.md**: Technical architecture and implementation approach
- **data-model.md**: Entity definitions and validation rules
- **tasks.md**: 46 completed tasks across 6 phases
- **research.md**: Technical decision documentation
- **contracts/api.rs**: Contract-first API definitions

## Quality Checks

- ✅ All tests passing (92/92)
- ✅ Zero clippy warnings
- ✅ CodeRabbit review issues resolved
- ✅ Documentation internally consistent
- ✅ Security requirements verified

## Related Work

- Builds on **WI-3** (heading extraction)
- Prepares for **WI-5** (section association via source line numbers)
- Defers compound statement splitting to **WI-6**
- Defers normative language detection to **WI-33**

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added extraction of structured content from Markdown, including ordered/unordered lists, tables, and paragraphs with source line tracking and nesting depth support.

* **Documentation**
  * Added comprehensive specification documents including feature requirements, data model definitions, implementation plan, and research notes for new structural extraction capability.

### CODE REVIEW

File: src/parse/clauses.rs [hunk 1]
```diff
@@ -0,0 +1,1000 @@
...
+            Event::SoftBreak if in_standalone_paragraph => para_text.push(' '),
+
+            // Ignore all other events (formatting tags, etc.)
+            _ => {}
+        }
+    }
+
+    // Sort all items by source_line to maintain document order
+    list_items.sort_by_key(|item| item.source_line);
+
```
Comment (brianluby) [inline src/parse/clauses.rs]: **Regarding O(n log n) complexity for sorting:**

The current implementation sorts list items at the end (O(n log n)) rather than maintaining document order during parsing. While this technically violates the "single-pass O(n)" claim in the algorithm documentation, this design is acceptable for the following reasons:

1. **Correctness First**: The sorting is necessary for correctness. Nested list items complete (and are pushed) when their  events fire, which means deeply nested children complete before their parent items, causing them to appear out of document order in the results vector.

2. **Practical Performance**: The number of list items in a policy document is typically small (tens to hundreds, not thousands). For these scales, the O(n log n) factor is negligible compared to the O(n) markdown parsing itself.

3. **Implementation Simplicity**: Maintaining document order during parsing would require a more complex data structure (e.g., a stack of indices with item slots pre-allocated), which would increase code complexity without meaningful performance benefit for the expected workload.

4. **Memory Locality**: The current approach benefits from better cache locality since all items are in a contiguous vector before sorting.

If we find this becomes a bottleneck in practice (which seems unlikely given policy document sizes), we can optimize later with profiling data. For now, the simple, correct implementation is preferred over premature optimization.

@Copilot
